### PR TITLE
Implement parser and typescript code generator

### PR DESCRIPTION
**Pull request by osipov-mit**

### CODE REVIEW

**Comment by NikVolf** [comment]
I would've called directory `js`, `sails-js` is redundant

File: js/src/generate/service-gen.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+              this._out
+                .line(`const replyPayloadBytes = await this.submitMsgAndWaitForReply(`, false)
+                .increaseIndent()
+                .line('this.programId,', false)
+                .line('payload,', false)
+                .line('account,', false)
+                .reduceIndent()
+                .line(')')
+                .line(`const result = this.registry.createType('${getType(output, true)}', replyPayloadBytes)`)
+                .line(`return result.toJSON() as ${returnType}`);
```
Comment (nikitayutanov) [inline js/src/generate/service-gen.ts:85]: `.toJSON`, as we've encountered in the past, is parsing some numeric types to hex. Wouldn't this be a problem?

**Review by nikitayutanov — APPROVED** [review]

File: js/src/generate/service-gen.ts (hunk 1 above)
Comment (osipov-mit) [reply js/src/generate/service-gen.ts]: Refactored. Will return number or bigint in case of numeric types and string in case of string type

**Review by DennisInSky — COMMENTED** [review]
I've left some comments along the code, but actually the main comment is that didn't we want to have our parser to be a wasm allowing it to be re-used from other languages? As such wouldn't it be better to use rust for implementing it then? I checked candid and they use `logos` as lexer in conjunction with `larlpop` as parser for building AST
@NikVolf , @osipov-mit

File: js/.gitignore
[the remark this answers is not in the document]
Comment (osipov-mit) [reply js/.gitignore:1]: removed

File: .github/workflows/sails-js.yaml
[the remark this answers is not in the document]
Comment (osipov-mit) [reply .github/workflows/sails-js.yaml:1]: done

File: .github/workflows/sails-js.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
+name: 'CI-CD sails-js'
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, labeled]
+    branches: [master]
+  push:
+    branches: [master]
+    paths:
+      - js/**
```
[the remark this answers is not in the document]
Comment (osipov-mit) [reply .github/workflows/sails-js.yaml]: Should be done in the future

**Pull request merged**